### PR TITLE
Raveena/wizard stepper list

### PR DESCRIPTION
**Pull request by raveenakothapally**
PLEASE SQUASH

Issue #191

Screen Shot 2020-03-11 at 1 31 40 AM

Screen Shot 2020-03-11 at 1 32 33 AM

Screen Shot 2020-03-11 at 1 32 44 AM

### CODE REVIEW

**Comment by ianmcburnie** [comment]
@raveenakothapally Some conflicts in `docs.min.js`. I think you might just need to re-run yarn then run a build.

**Comment by raveenakothapally** [comment]
> BEM

Thanks for the review @ianmcburnie,

A lot of BEM :) Let me read through some of these rules, your suggestions and then update the selectors accordingly.

File: src/less/wizard-stepper/base/wizard-stepper.less [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+.wizard-stepper-list__item--confirmation,
+.wizard-stepper-list__item--inprogress {
+    .wizard-stepper-list__item__node::before {
+        background-color: @wizard-step-confirmation-color;
+    }
+}
+
+.wizard-stepper-list__item__title {
```
Comment (ianmcburnie) [inline src/less/wizard-stepper/base/wizard-stepper.less]: BEM only goes as far as children, not grandchildren. So this would have to be `.wizard-stepper-list__item-title`. I know, it kind of feels like cheating, since title belongs to the item! But this is the way we have to do it.

File: src/less/wizard-stepper/base/wizard-stepper.less [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    flex-direction: column;
+    position: relative;
+    width: 30px;
+
+    &::before {
+        background-color: @wizard-step-default-color;
+        bottom: 35px;
+        content: '';
+        height: 3px;
+        left: -50px;
```
Comment (agliga) [inline src/less/wizard-stepper/base/wizard-stepper.less]: I have a bit of a problem with this, when I looked on firefox it not aligned correctly when I loaded it initially.
image


Here's my take, it might take a bit of a change but hopefully not too much.

The item__node should be flex and direction row. Afterwards the svg and text should be in it's own container.
The `:before` should have no position absolute, no left, no bottom. The flex will handle it correctly and the container.
The container with the SVG and text should have display:flex and flex-direction column. This way we let flex handle the `---` instead of doing absolute positioning which will never look correct. 

It's a bit complex. I played around with this locally and I was able to get something decent using just flex. It needs some slight work, but here's the diff that I was able to get working. (I had to upload it as a zip)
[patch.diff.zip](https://github.com/eBay/skin/files/4232538/patch.diff.zip)
You need some color changes and minor markup changes. As well as some max width. But this should be working mostly. It also aligns the text and arrows correctly (hopefully)
 
It still needs some size adjustments I believe as well as a max-width or something

However this is also the case with this version as well.
image

File: docs/_includes/common/wizard-stepper.html [hunk 3]
```diff
@@ -1,72 +1,58 @@
...
-            <div class="wizard-step-container--confirmation">
-                <div class="wizard-step">
-                    <svg class="icon icon--confirmation-filled" focusable="false" height="24" width="24">
-                        <use xlink:href="#icon-confirmation-filled"></use>
-                    </svg>
-                    <span class="wizard-step-title">
-                        Started
-                    </span>
-                </div>
+    <ul class="wizard-stepper-list">
```
Comment (ianmcburnie) [inline docs/_includes/common/wizard-stepper.html]: There can be only one base class in a module. I currently see two: `wizard-stepper` and `wizard-stepper-list`. So we have to decide on one.

The reason I wanted to go with `wizard-stepper-list` (which is quite long, I know) was to differentiate it from the other type of wizard stepper that uses cards. However... I now think that we can just use `wizard-stepper` as the base class for this version you are working on here. It feels like this version will be the most often used one, and so we can consider it the default.

This would mean changing:

* `wizard-stepper-list` to `wizard-stepper__items`
* `wizard-stepper-list__item--confirmation` to `wizard-stepper__item wizard-stepper__item--confirmation`
* etc

Basically removing the word list from everywhere. Happy to wait and go through this in person if it would make more sense.

I also had an idea where we might be able to use this same structure for the other type of wizard stepper and drop card-type steps into the list, rather than icon-and-line-type steps. But let's get this version done first and see if that would work later... :-)

Comment (raveenakothapally) [reply docs/_includes/common/wizard-stepper.html]: @ianmcburnie I do not have a class called `wizard-stepper` which you are referring as another base class. My base class is `wizard-stepper-list` which I thought would go hand in hand with HTML list (ul, ol) and list items (li)

Comment (ianmcburnie) [reply docs/_includes/common/wizard-stepper.html]: Oh my bad, it's an ID not a class. Let me re-review this aspect...

**Comment by ianmcburnie** [comment]
@raveenakothapally I'm running into a few lint errors that will need to be fixed:

Screen Shot 2020-03-02 at 9 37 49 AM

File: docs/_includes/common/wizard-stepper.html (hunk 3 above)
Comment (ianmcburnie) [reply docs/_includes/common/wizard-stepper.html]: I guess I got confused because the example is missing the usual containing structure of `<div class="demo"><div class="demo__inner">..</div></div>`. Could you please add that? Adding the example markup right afterwards would be useful too.

Comment (ianmcburnie) [reply docs/_includes/common/wizard-stepper.html]: So I'm thinking something like this would get us closer to our ideal HTML structure and BEM naming:

```
<div class="wizard-stepper">
    <ul class="wizard-stepper__items">
        <li class="wizard-stepper__item wizard-stepper__item--confirmation">
            <span class="wizard-stepper__icon">
                <svg class="icon icon--confirmation-filled" focusable="false" height="24" width="24">
                    <use xlink:href="#icon-confirmation-filled"></use>
                </svg>
            </span>
            <span class="wizard-stepper__text">Started</span>
        </li>
        ..
    </ul>
</div>
```

Vertical might only need a base class modifier:

```
<div class="wizard-stepper wizard-stepper--vertical">
    <!-- same as above example -->
</div>
```

We could possibly use a similar structure for the card/form based stepper (when we get to it later):

```
<div class="wizard-stepper">
    <ul class="wizard-stepper__cards">
        <li class="wizard-stepper__card wizard-stepper__card--confirmation">
            ..
        </li>
        ..
    </ul>
</div>
```

Although we may end up only just needing something as simple as this, with no list structure, we shall see:

```
<div class="wizard-card">
    ..
</div>
```

Let's discuss further in our meeting today.

**Comment by raveenakothapally** [comment]
> @raveenakothapally This is looking great! Very close to the finish line I think. One last thing I think needed is an example of how the lines will "grow", for example here they stretch vertically when more content is added:
> 
> Screen Shot 2020-03-12 at 8 48 29 AM
> 
> It would be great to also show that the skin module can be "fluid" and fill all available space. We would probably use a `wizard-stepper--fluid` modifier for this, like how we do in other modules. You can see in the screenshot below that a custom width has been set on the wizard stepper. A wizard stepper with fluid modifier would grow to fill that width.
> 
> Screen Shot 2020-03-12 at 8 48 42 AM

hello @ianmcburnie 
I would use the `wizard-stepper--fluid` class on `ul`. This class then makes all its `li` do `flex: 1`.
However, it throws off the line between the nodes to not expand because of existing fixed `width`.
Theoretically, I will see if I can expand the line to 100% and let the nodes hide the lines that might have overlapped with the lines. I am experimenting these ideas now.

Please let me know if you have other ideas.

File: src/svg/ds6/icons.svg [hunk 4]
```diff
@@ -102,8 +102,19 @@
     <symbol viewBox="0 0 8 14" id="icon-chevron-right-small">
         <path d="M1.005 14a1 1 0 01-.71-1.71l5.296-5.288L.296 1.714A1.004 1.004 0 011.714.294L7.71 6.292a1 1 0 010 1.41L1.714 13.7a.999.999 0 01-.709.3z"></path>
     </symbol>
+<<<<<<< HEAD
```
Comment (ianmcburnie) [inline src/svg/ds6/icons.svg]: Looks like some leftovers from a merge conflict.

File: src/less/wizard-stepper/base/wizard-stepper.less [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+.wizard-stepper--fluid {
+    & > .wizard-stepper__item {
+        flex: 1;
+        max-width: 220px;
+    }
+}
+
+.wizard-stepper.wizard-stepper--vertical {
```
Comment (ianmcburnie) [inline src/less/wizard-stepper/base/wizard-stepper.less]: Do we need the extra specificity of a chained class here? Ideally would just be `.wizard-stepper--vertical`.

File: docs/_includes/tests/wizard-stepper.html [hunk 6]
```diff
@@ -0,0 +1,125 @@
+<h3 id="wizard-stepper-default">Default Wizard Stepper</h3>
```
Comment (ianmcburnie) [inline docs/_includes/tests/wizard-stepper.html]: I think we can drop these test files now. Storybook is the way forward.

File: docs/_includes/common/wizard-stepper.html [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+    </ul>
+</div>
+    {% endhighlight %}
+
+    <h3 id="wizard-stepper-vertical">Fluid Wizard Stepper</h3>
+    <p>
+        Stepper width will increase as the screen size stretches with a maximum total individual step reaching 220px.
+    </p>
+    <p>
+        Just add <em>wizard-stepper--fluid</em> class to the <em>ul.wizard-stepper__items</em> tag.
```
Comment (ianmcburnie) [inline docs/_includes/common/wizard-stepper.html]: For correct BEM, the class should be `wizard-stepper__items--fluid`.

File: dist/wizard-stepper/ds4/wizard-stepper.css [hunk 8]
```diff
@@ -0,0 +1,206 @@
+.wizard-stepper__items {
+  -webkit-box-align: center;
+          align-items: center;
+  display: -webkit-box;
+  display: flex;
+  flex-wrap: wrap;
+  -webkit-box-pack: center;
+          justify-content: center;
```
Comment (ianmcburnie) [inline dist/wizard-stepper/ds4/wizard-stepper.css]: Been thinking about this center alignment, and after looking through the mocks, I think we should have left aligned by default, and center aligned as an opt-in modifier, e.g. `wizard-stepper__items--center`.

**Comment by ianmcburnie** [comment]
@raveenakothapally I'm running into this issue when I run `yarn start`:

Screen Shot 2020-03-27 at 1 47 32 PM

Maybe some old files or references left over in your branch?

**Comment by raveenakothapally** [comment]
> @raveenakothapally I'm running into this issue when I run `yarn start`:
> 
> Screen Shot 2020-03-27 at 1 47 32 PM
> 
> Maybe some old files or references left over in your branch?

Hello Ian, looks like it. I force refreshed the branch. Can you try now?
